db/clientlist_virtual_reader.cc

```diff
@@ -0,0 +1,237 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "db/clientlist_virtual_reader.hh"
+
+#include "api/api.hh"
+#include "db/system_keyspace.hh"
+#include "db/timeout_clock.hh"
+#include "dht/i_partitioner.hh"
+#include "mutation_fragment.hh"
+#include "range.hh"
+#include "service/storage_service.hh"
+#include "transport/server.hh"
+
+#include <algorithm>
+#include <unordered_map>
+#include <vector>
+
+namespace db::clientlist {
+
+namespace {
+class clients_reducer {
+    std::vector<system_keyspace::client_data> _val;
+public:
+    future<> operator()(std::vector<system_keyspace::client_data>&& rhs) {
+        // Move the smaller vector into the bigger one
+        if (_val.size() < rhs.size()) {
+            _val.swap(rhs);
+        }
+        _val.reserve(_val.size() + rhs.size());
+        _val.insert(_val.end(), std::make_move_iterator(rhs.begin()), std::make_move_iterator(rhs.end()));
+        return make_ready_future<>();
+    }
+
+    std::vector<system_keyspace::client_data> get() && {
+        return std::move(_val);
+    }
+};
+
+class ck_comparator {
+    schema_ptr _schema;
+public:
+    using port_type = decltype(system_keyspace::client_data::port);
+
+    ck_comparator(schema_ptr schema) : _schema(schema) {}
+
+    int operator()(const clustering_key_prefix& lhs, const clustering_key_prefix& rhs) {
+        return clustering_key_prefix::prefix_equal_tri_compare(*_schema)(lhs, rhs);
+    }
+    int operator()(const port_type& lhs, const clustering_key_prefix& rhs) {
+        const auto lhs_bytes = data_type_for<port_type>()->decompose(lhs);
+        return operator()(clustering_key_prefix::from_single_value(*_schema, lhs_bytes), rhs);
+    }
+    int operator()(const clustering_key_prefix& lhs, const port_type& rhs) {
+        const auto rhs_bytes = data_type_for<port_type>()->decompose(rhs);
+        return operator()(lhs, clustering_key_prefix::from_single_value(*_schema, rhs_bytes));
+    }
+};
+}  // anon. namespace
+
+static future<std::vector<system_keyspace::client_data>> get_all_clients_data(
+        distributed<cql_transport::cql_server>& distributed_cql_server) {
+    return distributed_cql_server.map_reduce(clients_reducer{},
+            [] (const cql_transport::cql_server& srv) {
+        const auto& conn_list = srv.get_connections_list();
+        return do_with(std::vector<system_keyspace::client_data>{},
+                [&conn_list] (auto& per_shard_result) {
+            return do_for_each(conn_list.begin(), conn_list.end(),
```

Comment:
Isn't it possible for the _connections_list inside cql_server to change in-between tasks?
In that case these iterators could be invalidated while `do_for_each` is running.

Reply:
It would be nice to have a dtest which creates a lot of connections and concurrently queries this table.

Reply:
You can just change this to make a copy of the conn list to make it safe.